### PR TITLE
Analysis refactor

### PR DESCRIPTION
**Pull request by somu-imply**
1. The current data source analysis involves a lot of instanceof checks and with the addition of future datasources is going to be messy. Moreover these are the issues found with the analysis:

2. The top level datasource is always known when the analysis is created so there is no need for that to be present in analysis. Actually analysis.getDataSource() is same as the datasource which created the analysis using the forDataSource call

3. Each datasource has a isGlobal check, no need for it to be a part of analysis

4. Complex instance of checks in the forDataSource method

5. isQuery check is basically checking if top level data source is an instance of QueryDataSource which can be taken care of by the callee.

This refactoring is to address these issues and move the analysis inside data source. A new method `getAnalysis()` is introduced in each datasource. For getting the base query a new method is added in DataSourceAnalysis that gets triggered through QueryDataSource for adding the baseQuery to the analysis only if there is no baseQuery present. This ensures that while moving up the tree, the lowermost baseQuery is stored in the DataSourceAnalysis object. The test cases have been updated to use these changes.

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/client/BrokerServerView.java [hunk 1]
```diff
@@ -353,7 +353,7 @@ public Optional<VersionedIntervalTimeline<String, ServerSelector>> getTimeline(f
   {
     final TableDataSource table =
         analysis.getBaseTableDataSource()
-                .orElseThrow(() -> new ISE("Cannot handle datasource: %s", analysis.getDataSource()));
+                .orElseThrow(() -> new ISE("Cannot handle datasource: %s", analysis.getBaseDataSource()));
```
[the remark this answers is not in the document]
Comment (somu-imply) [reply server/src/main/java/org/apache/druid/client/BrokerServerView.java:356]: This is interpolating the base data source into the message. Maybe I need to update the message as the original analysis.getDataSource() does not exist anymore

File: processing/src/main/java/org/apache/druid/query/DataSource.java [hunk 2]
```diff
@@ -124,4 +124,7 @@
    */
   byte[] getCacheKey();
 
+  DataSourceAnalysis getAnalysisForDataSource();
+
+  DataSourceAnalysis getAnalysisForDataSource(Query<?> query);
```
[the remark this answers is not in the document]
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/DataSource.java]: Addressed, query data source now populates the baseQuery while moving up the tree. A method to update with new analysis object has been added

File: processing/src/main/java/org/apache/druid/query/QueryDataSource.java [hunk 3]
```diff
@@ -112,6 +113,25 @@ public byte[] getCacheKey()
...
+    return getAnalysisForDataSource(null);
+  }
+
+  @Override
+  public DataSourceAnalysis getAnalysisForDataSource(Query<?> query)
+  {
+    final Query<?> subQuery = this.getQuery();
+    if (!(subQuery instanceof BaseQuery)) {
+      // We must verify that the subQuery is a BaseQuery, because it is required to make "getBaseQuerySegmentSpec"
+      // work properly. All built-in query types are BaseQuery, so we only expect this with funky extension queries.
```
[the remark this answers is not in the document]
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/QueryDataSource.java]: The comment has been updated to point to `DataSourceAnalysis. getBaseQuerySegmentSpec`

File: processing/src/main/java/org/apache/druid/query/QueryDataSource.java [hunk 4]
```diff
@@ -112,6 +113,25 @@ public byte[] getCacheKey()
...
+      // We must verify that the subQuery is a BaseQuery, because it is required to make
+      // "DataSourceAnalysis.getBaseQuerySegmentSpec" work properly.
+      // All built-in query types are BaseQuery, so we only expect this with funky extension queries.
+      throw new IAE("Cannot analyze subquery of class[%s]", subQuery.getClass().getName());
+    }
+    final DataSource current = subQuery.getDataSource();
+    DataSourceAnalysis currentAnalysis = current.getAnalysisForDataSource();
+    // check if the baseQuery is already present while moving up
+    // if not add the current query as the baseQuery to the analysis
+    if (!currentAnalysis.getBaseQuery().isPresent()) {
```
[the remark this answers is not in the document]
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/QueryDataSource.java]: Great idea. Done

File: processing/src/main/java/org/apache/druid/query/JoinDataSource.java [hunk 5]
```diff
@@ -479,4 +481,47 @@ public byte[] getCacheKey()
     }
     return keyBuilder.build();
   }
+
+  @Override
+  public DataSourceAnalysis getAnalysisForDataSource()
```
[the remark this answers is not in the document]
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/JoinDataSource.java]: The one in the constructor was calling this only. Made a slight change to have the analysis in the constructor call a private method to populate the analysis and this public method just to return the one in the constructor

File: processing/src/main/java/org/apache/druid/query/JoinDataSource.java [hunk 6]
```diff
@@ -482,13 +482,18 @@ public byte[] getCacheKey()
...
-  @Override
-  public DataSourceAnalysis getAnalysisForDataSource()
+  private DataSourceAnalysis getAnalysis()
   {
     final Triple<DataSource, DimFilter, List<PreJoinableClause>> flattened = flattenJoin(this);
     return new DataSourceAnalysis(flattened.first, null, flattened.second, flattened.third);
   }
 
+  @Override
+  public DataSourceAnalysis getAnalysisForDataSource()
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/JoinDataSource.java:492]: This rename made me realize...  I like the name `getAnalysis()` a lot more for the method on the interface.  Can we rename that?

**Review by imply-cheddar — COMMENTED** [review]
One last comment.

File: processing/src/main/java/org/apache/druid/query/JoinDataSource.java (hunk 6 above)
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/query/JoinDataSource.java]: Done :)

**Comment by somu-imply** [comment]
The branch coverage is failing on me for server module although my refactor just renamed a couple of things there

**Comment by imply-cheddar** [comment]
It's complaining that parts of the if conditions that you changed when you adjusted the `isQuery()` are not covered in the tests.  It needs a branch coverage of 50%, currently 11 of 24 branches are covered.  So you just need one more branch covered.  Just for posterity, it's probably great to try to find a way to cover at least one of those branches...

It looks like there is a `ServerManagerTest` and a `ClientQuerySegmentWalkerTest`.  It is probably worth it to just look and see if there's a relatively straightforward way to build tests for one/both of those classes that get the exception to be thrown.

Other than that, i think this is ready for merge.

**Pull request merged**